File: trip-kit-booking/src/main/java/com/skedgo/android/tripkit/booking/BookingError.java
package com.skedgo.android.tripkit.booking;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.Nullable;

import com.google.gson.annotations.SerializedName;

public class BookingError extends Throwable {

  @Nullable @SerializedName("title")
  private String title;
  @SerializedName("errorCode")
  private int errorCode;
  @SerializedName("error")
  private String error;
  @SerializedName("usererror")
  private boolean hasUserError;
  @Nullable @SerializedName("recovery")
  private String recovery;
  @Nullable @SerializedName("recoveryTitle")
  private String recoveryTitle;
  @Nullable @SerializedName("url")
  private String url;

  @Nullable
  public String getTitle() {
    return title;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  public int getErrorCode() {
    return errorCode;
  }

  public void setErrorCode(int errorCode) {
    this.errorCode = errorCode;
  }

  public String getError() {
    return error;
  }

  public void setError(String error) {
    this.error = error;
  }

  @Nullable public String getRecovery() {
    return recovery;
  }

  public void setRecovery(@Nullable String recovery) {
    this.recovery = recovery;
  }

  @Nullable public String getRecoveryTitle() {
    return recoveryTitle;
  }

  public void setRecoveryTitle(@Nullable String recoveryTitle) {
    this.recoveryTitle = recoveryTitle;
  }

  @Nullable public String getUrl() {
    return url;
  }

  public void setUrl(@Nullable String url) {
    this.url = url;
  }

  public boolean hasUserError() {
    return hasUserError;
  }

  public void setHasUserError(boolean hasUserError) {
    this.hasUserError = hasUserError;
  }

  @Override public String getMessage() {
    if (error != null) {
      return error;
    } else {
      return super.getMessage();
    }
  }

}

File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/BookingError.kt
package com.skedgo.tripkit.booking

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName

data class BookingErrorData(
    @SerializedName("title")
    var title: String? = null,

    @SerializedName("errorCode")
    var errorCode: Int = 0,

    @SerializedName("error")
    var error: String? = null,

    @SerializedName("usererror")
    var hasUserError: Boolean = false,

    @SerializedName("recovery")
    var recovery: String? = null,

    @SerializedName("recoveryTitle")
    var recoveryTitle: String? = null,

    @SerializedName("url")
    var url: String? = null
)

// Custom Exception class that takes a JSON string as input
class BookingError(val errorData: BookingErrorData) : Throwable(errorData.error) {
    companion object {
        // Factory method to parse JSON and create the exception
        fun fromJson(jsonData: String): BookingError {
            val errorData = Gson().fromJson(jsonData, BookingErrorData::class.java)
            return BookingError(errorData)
        }
    }

    val title: String?
        get() {
            return errorData.title
        }

    val errorCode: Int
        get() {
            return errorData.errorCode
        }

    val error: String?
        get() {
            return errorData.error
        }

    val hasUserError: Boolean
        get() {
            return errorData.hasUserError
        }

    val recovery: String?
        get() {
            return errorData.recovery
        }

    val recoveryTitle: String?
        get() {
            return errorData.recoveryTitle
        }

    val url: String?
        get() {
            return errorData.url
        }
}
